#! /bin/bash

#SBATCH  --job-name=star_solo
#SBATCH --mail-type=ALL
#SBATCH --mail-user j.chouaref@lumc.nl
#SBATCH -t 1:0:0
#SBATCH --mem=60000

#/path/to/STAR --genomeDir /path/to/genome/dir/ --readFilesIn ...  [...other parameters...] --soloType ... --soloCBwhitelist ...

READS="/exports/humgen/jihed/scTE/pbmc"
INDEX="/exports/humgen/jihed/TE_Transcript/temp/genome"

STAR --genomeDir $INDEX \
    --readFilesIn $READS/subset_pbmc_1k_v3_S1_L001_R2_001.fastq.gz $READS/subset_pbmc_1k_v3_S1_L001_R1_001.fastq.gz \
    --soloUMIlen 12 --soloType CB_UMI_Simple --soloCBwhitelist 3M-february-2018.txt 

STAR --genomeDir $INDEX \
    --readFilesIn $READS/subset_pbmc_1k_v3_S1_L002_R1_001.fastq.gz $READS/subset_pbmc_1k_v3_S1_L002_R2_001.fastq.gz \
    --soloUMIlen 12 --soloType CB_UMI_Simple --soloCBwhitelist 3M-february-2018.txt 

#For bam file generated by STARsolo etc, the cell barcodes and UMI need to be integrated into the read 'CR:Z' or 'UR:Z' tags as bellow: