{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test of the output of the new RNA-seq analysis snakemake workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HTseq output\n",
    "\n",
    "The workflow produces a count table obtained with HTSEQ named `htseq_genes.txt` (tab separated table). You can find the count table in the folder: `results/htseq/genes`\n",
    "\n",
    "It should look like this: \n",
    "\n",
    "`less htseq_genes.txt` \n",
    "\n",
    "```\n",
    "ENSMUST00000000001.4    3796    2340    2380    1157    2312    2075    2430    2144    2311    2454    2255\n",
    "ENSMUST00000000003.13   0       0       0       0       0       0       0       0       0       0       0\n",
    "ENSMUST00000000010.8    1       0       0       0       0       0       0       3       3       0       0\n",
    "ENSMUST00000000028.13   0       0       0       0       0       0       0       0       0       0       0\n",
    "ENSMUST00000000033.11   0       0       0       0       0       0       0       0       0       0       0\n",
    "ENSMUST00000000049.5    0       0       0       0       1       0       0       0       0       0       0\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup the Docker image to perform the analysis on your own computer\n",
    "\n",
    "To perform the analysis on your own computer, you will need the administartors rights. \n",
    "\n",
    "- Install Docker on your computer by downloading it the version corresponsing to your [OS](https://docs.docker.com/get-docker/). Follow the instructions for the installation. \n",
    "\n",
    "- Once the installation is done, start `Docker Desktop`\n",
    "\n",
    "- Prepare a folder (refered from now as `working directory` where you will perform the analysis. The analysis requires 2 files:\n",
    "    - The count table `htseq_genes.txt`.\n",
    "    - The `units.tsv` you used for the analysis.\n",
    "\n",
    "- Then using your usual Terminal (I recommand [Visual code studio](https://code.visualstudio.com/download)), open a terminal. \n",
    "\n",
    "- Change directory in order to get to the `working directory`\n",
    "\n",
    "- Set up the variable `WD` with the following command:\n",
    "\n",
    "```\n",
    "WD=pwd\n",
    "``` \n",
    "\n",
    "- Check that the variable is set correctly with the following command: `$WD`. This command should return the absolute path of the `working directory`. \n",
    "\n",
    "- Initiate Docker with: \n",
    "\n",
    "```\n",
    "docker run --rm -it -d -v $WD  jchouaref/bioconductor_biomart:1.0\n",
    "```\n",
    "\n",
    "If it initate correclty then, the name of the image container should appear in the terminal: `ab79cc2dfb22f01b9198706bf46c2faa02c6814a5dfcd4d570844a963ef3b889`\n",
    "\n",
    "- You should be then able to start a R session by typing: `R` "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform the analysis in R \n",
    "\n",
    "Once the R session is initated, start by initiating the variable we will use for the analysis with the following:\n",
    "```\n",
    "count_df <- \"featureCounts_genes.txt\"\n",
    "file_experiment <- \"units.tsv\"\n",
    "```\n",
    "\n",
    "Write here either the absolute or relative path to the output folder where the output of DESeq2 will be saved: \n",
    "\n",
    "```\n",
    "output <- \"results/\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the required R packages:\n",
    "\n",
    "```\n",
    "library(ggplot2)\n",
    "library(DESeq2)\n",
    "library(\"RColorBrewer\")\n",
    "library(pheatmap)\n",
    "library(\"biomaRt\")\n",
    "library(Glimma)\n",
    "library(EnhancedVolcano)\n",
    "library(readxl)\n",
    "library(dplyr)\n",
    "library(readr)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare the count table using the `units.tsv` file \n",
    "\n",
    "The units.tsv file contains the files names and the conditions, you will need to use for the analysis.\n",
    "\n",
    "- First set up the condition variables:\n",
    "\n",
    "```\n",
    "condition1 <- \"treatment\"\n",
    "condition2 <- \"control\"\n",
    "```\n",
    "\n",
    "The conditions need to have the same name as what you defined in the `units.tsv` file.\n",
    "\n",
    "- Load the `units.tsv` table and save it in the variable: `design_exp`: `design_exp <- read.table(file_experiment, header = TRUE, row.names = \"sample\")`\n",
    "\n",
    "- Then load the count table and set the header names :\n",
    "    - First load the count table with the following:\n",
    "    ```\n",
    "    count_table <- read_delim(count_df,\n",
    "    delim = \"\\t\",  escape_double = FALSE,\n",
    "    trim_ws = TRUE)\n",
    "    ```\n",
    "\n",
    "    The count table might not have headers. Headers are necessary for the script.\n",
    "    They also need to have the same name as the samples for the `design_exp` table\n",
    "    You can set names for the headers for this file.\n",
    "\n",
    "    - Set the header to the `count_table` with the following:\n",
    "    ``` \n",
    "    colnames(count_table) <- c(\"Gene\", \"RNA1\", \"RNA2\", \"RNA3\", \"RNA4\", \"RNA5\", \"RNA6\")\n",
    "    ```\n",
    "The column headers are in the **exact same order** as in the `units.tsv` file. If there is a typo in the names or a change in their order, R will raise a error message.\n",
    "\n",
    "- Prepare count table to perform the DESeq2 aanalysis with the following commands:\n",
    "\n",
    "```\n",
    "row.names(count_table) <- count_table$Gene\n",
    "\n",
    "count_table_df <- count_table[, 2:ncol(count_table)]\n",
    "\n",
    "row.names(count_table_df) <- count_table$Gene\n",
    "\n",
    "design_exp <- design_exp[names(count_table_df), ]\n",
    "\n",
    "design_exp$condition <- factor(design_exp$condition)\n",
    "\n",
    "```\n",
    "\n",
    "### Running the DESeq2 analysis\n",
    "\n",
    "```\n",
    "dds <- DESeqDataSetFromMatrix(countData = count_table_df,\n",
    "                            colData = design_exp,\n",
    "                            design = ~condition)\n",
    "\n",
    "dds$condition <- relevel(dds$condition, ref = \"control\")\n",
    "\n",
    "dds <- DESeq(dds)\n",
    "\n",
    "res2 <- results(dds, contrast = c(\"condition\", condition1, condition2 ))\n",
    "\n",
    "rld = rlog(dds, blind = TRUE) # nolint\n",
    "\n",
    "```\n",
    "\n",
    "#### Save output table\n",
    "\n",
    "```\n",
    "write.table(res2,\n",
    "    file = paste0(output, \"deseq2_table_\", condition1, \"_vs_\", condition2, \".txt\"),\n",
    "    sep = \"\\t\",\n",
    "    quote = FALSE)\n",
    "```\n",
    "\n",
    "This table can be read using Excel and already contain all the information needed to know if there are differentially expressed genes. However the gene names are ensembl gene ID. In the next steps, we will transform this table to add the gene symbols to it as well as the read count and normalized read count per gene for each sample.\n",
    "\n",
    "#### PDF with QC measurments: PCA, correlation heatmap\n",
    "\n",
    "```\n",
    "pdf(file = paste0(output, \"MAplot_DGE_\", condition1, \"_vs_\", condition2, \".pdf\"),\n",
    "    paper = \"a4\")\n",
    "plotMA(res2)\n",
    "rld <- rlog(dds, blind = TRUE)\n",
    "plotPCA(rld,\n",
    "    n = 10000)\n",
    "\n",
    "sample_dists <- dist(t(assay(rld)))\n",
    "sample_dist_matrix <- as.matrix(sample_dists)\n",
    "rownames(sample_dist_matrix) <- rld$condition\n",
    "colnames(sample_dist_matrix) <- NULL\n",
    "colors <- colorRampPalette(rev(brewer.pal(9, \"Blues\")))(255)\n",
    "\n",
    "pheatmap(sample_dist_matrix,\n",
    "    clustering_distance_rows = sample_dists,\n",
    "    clustering_distance_cols = sample_dists,\n",
    "    col = colors)\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "#### Biomart annotation\n",
    "\n",
    "```\n",
    "genes   <- as.character(row.names(count_table_df))\n",
    "fixed_names = sapply(strsplit(genes, \".\", fixed=T), function(x) x[1])\n",
    "\n",
    "mart    <- useDataset(\"mmusculus_gene_ensembl\", useMart(\"ensembl\"))\n",
    "\n",
    "G_list  <- getBM(filters = \"ensembl_gene_id\",\n",
    "            attributes = c(\"ensembl_gene_id\",'mgi_symbol', 'chromosome_name',\n",
    "            'start_position', 'end_position', 'strand',\n",
    "            \"description\"),\n",
    "            values = fixed_names,\n",
    "            mart = mart)\n",
    "\n",
    "annotation <- aggregate(G_list[, \"mgi_symbol\"],\n",
    "    by = list(G_list$ensembl_gene_id, G_list$chromosome_name,\n",
    "            G_list$start_position, G_list$end_position,\n",
    "            G_list$strand,\n",
    "            G_list$description),\n",
    "            function(x){paste(x, sep=\"_\")})\n",
    "\n",
    "annotation <- aggregate(mgi_symbol~ensembl_gene_id+chromosome_name+start_position+end_position+strand+description,\n",
    "    data=G_list,\n",
    "    FUN = paste,\n",
    "    collapse = \",\")\n",
    "\n",
    "row.names(annotation) <- annotation$ensembl_gene_id\n",
    "\n",
    "annotation$name <- unlist(annotation$mgi_symbol)\n",
    "```\n",
    "\n",
    "#### Glimma\n",
    "\n",
    "```\n",
    "genes   <- as.character(row.names(res2))\n",
    "fixed_names = sapply(strsplit(genes, \".\", fixed=T), function(x) x[1])\n",
    "\n",
    "annotation <- annotation[fixed_names, ]\n",
    "\n",
    "annotation$Gene <- annotation$ensembl_gene_id\n",
    "\n",
    "dt.res2 <- as.numeric(res2$padj < 0.05)\n",
    "\n",
    "row.names(res2) <- fixed_names\n",
    "\n",
    "finalTable <- merge(annotation,\n",
    "    as.data.frame(res2),\n",
    "    by.x = \"row.names\",\n",
    "    by.y = \"row.names\")\n",
    "\n",
    "row.names(finalTable) <- finalTable$Row.names\n",
    "\n",
    "finalTable$Row.names <- c()\n",
    "\n",
    "dds_count <- counts(dds, normalized = FALSE)\n",
    "genes   <- as.character(row.names(dds_count))\n",
    "fixed_names = sapply(strsplit(genes, \".\", fixed=T), function(x) x[1])\n",
    "row.names(dds_count) <- fixed_names\n",
    "\n",
    "finalTable <- merge(finalTable,\n",
    "    dds_count,\n",
    "    by.x = \"row.names\",\n",
    "    by.y = \"row.names\")\n",
    "\n",
    "row.names(finalTable) <- finalTable$Row.names\n",
    "\n",
    "finalTable$Row.names <- c()\n",
    "\n",
    "dds_count_norm <- counts(dds, normalized = TRUE)\n",
    "genes   <- as.character(row.names(dds_count_norm))\n",
    "fixed_names = sapply(strsplit(genes, \".\", fixed=T), function(x) x[1])\n",
    "row.names(dds_count_norm) <- fixed_names\n",
    "\n",
    "finalTableNorm <- merge(finalTable,\n",
    "    dds_count_norm,\n",
    "    by.x = \"row.names\",\n",
    "    by.y = \"row.names\",\n",
    "    suffixes = c(\"\", \".norm\"))\n",
    "\n",
    "``` \n",
    "The following command will open the browser to display a MDS plot. It's a plot that discplay the **Euclidian distances** between your samples. It helps to know if the biological replicates are close to one another and also have a first idea of how similar the 2 conditions tested are.\n",
    "\n",
    "```\n",
    "glMDSPlot(dds, groups = design_exp, launch = TRUE)\n",
    "```\n",
    "\n",
    "The following command will also open the browser to display a MA plot, the MA plot will the log fold change measured for every genes. The genes above 0 are more expressed in the `treatment` condition compared to the `control` condition , while the genes below the 0 aree less expressed in the `treatment` condition compared to the `control` condition.\n",
    "\n",
    "```\n",
    "glMDPlot(res2, status = dt.res2,\n",
    "    counts = counts(dds, normalized = TRUE),\n",
    "    groups = dds$condition,\n",
    "    transform = TRUE,\n",
    "    side.main = \"mgi_symbol\",\n",
    "    samples = colnames(dds),\n",
    "    anno = annotation[, c(9, 7, 6)],\n",
    "    launch = TRUE)\n",
    "```\n",
    "\n",
    "### Save the final table \n",
    "\n",
    "write.table(finalTableNorm,\n",
    "    file = paste0(output, \"DESEQ2_final_table_\", condition1, \"_vs_\", condition2, \".txt\"), \n",
    "    sep = \"\\t\", \n",
    "    quote = FALSE, \n",
    "    row.names = FALSE)\n",
    "\n",
    "\n",
    "This is the table, you will want to use in excel to make the filtering for the padj and the log2FC. Since it contains the information about the counts and normalized counts per genes, it can also be useful to make heatmaps and GSEA.\n",
    "\n",
    "### Enhanced volcano plot\n",
    "\n",
    "The following part will show how to prepare a volcano plot in R. Note that the final table can also be loaded directly in [VolcanoseR](https://huygens.science.uva.nl/VolcaNoseR/) to produce volcano plot very easily.\n",
    "\n",
    "```\n",
    "res3 <- finalTableNorm %>%\n",
    "  select(name, baseMean, log2FoldChange, lfcSE, stat, pvalue, padj)\n",
    "```\n",
    "\n",
    "The values below will need to be adapted to each volcanoplot:\n",
    "\n",
    "```\n",
    "min_x_lim <- -4\n",
    "max_x_lim <- 5\n",
    "min_y_lim <- 0\n",
    "max_y_lim <- 20\n",
    "title_plot <- \"Volcano plot DGE treatment vs control\"\n",
    "```\n",
    "\n",
    "```\n",
    "pdf(paste0(output, \"Enhanced_volcano_plot_\", condition1, \"_vs_\", condition2, \".pdf\"), height = 10, width = 10)\n",
    "  EnhancedVolcano(res3,\n",
    "    lab = res3$name,\n",
    "    subtitle = \"\",\n",
    "    x = \"log2FoldChange\",\n",
    "    y = \"padj\",\n",
    "    xlim = c(min_x_lim, max_x_lim),\n",
    "    ylim = c(min_y_lim, max_y_lim),\n",
    "    pCutoff = 0.05,\n",
    "    ylab = \"-Log10 padj value\",\n",
    "    title = title_plot,\n",
    "    )\n",
    "  \n",
    "dev.off()\n",
    "```\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.6 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
